accepted/future-releases/nnbd/number-operation-typing.md

```diff
@@ -62,47 +62,76 @@ T add<T extends num>(T a, T b) => a + b;
-> * Otherwise if *T* <: `int` and *S* <: `int` then the static type of `e` is `int`.
-> * Otherwise if *T* <: `double` or *S* <: *double* then the static type of *e* is `double`.
+> * If *T* <: `int` and *S* <: `int` then the static type of `e` is *T*.
+> * Otherwise if *T* <: `double` , then the static type of `e` is *T*.
+> * Otherwise if *S* <: `double` then the static type of `e` is `double`.
+> * Otherwise if *T* is not a subtype of `int` or `double`, then the static type of `e` is *T*.
 > * Otherwise the static type of *e* is `num`.
 
 We also special-case the `clamp` method:
 
 > For a normal invocation `e` of the form `e1.clamp(e2, e3)`, where the static types of `e1`, `e2` and `e3` are *T*<sub>1</sub>, *T*<sub>2</sub> and *T*<sub>3</sub> respectively, which are all non-`Never` subtypes of `num`:
 >
-> * If any of *T*<sub>1</sub>, *T*<sub>2</sub> or *T*<sub>3</sub> is a supertype of the other two, the first such is the static type of `e`.
+> * If any of *T*<sub>1</sub>, *T*<sub>2</sub> or *T*<sub>3</sub> is a supertype of the other two, the first such (in source order of the corresponding operands) is the static type of `e`.
 > * Otherwise, if all of *T*<sub>1</sub>, *T*<sub>2</sub> and *T*<sub>3</sub> are subtypes of `int`, the static type of `e` is `int`.
 > * Otherwise if all of *T*<sub>1</sub>, *T*<sub>2</sub> and *T*<sub>3</sub> are subtypes of `double`, the static type of `e` is `double`.
-> * Otherwise the static type of `e` is `num`.
+> * Otherwise the static type of `e` is `num`.
 
 With these extensions, we cover all members on `int` which has a return type of `num`, and we ensure that a using operands with the *same* type gives a result of that type, even if that type is a type variable (like `X extends int`) or promoted type variable (like `X & int`).
 
-There are no special rules for `/` and `~/` because their return type is not `num`, and the return value's type is independent of the argument types. A `/` operation always returns a `double` and a `~/` operation always returns an `int`.
+There are no special rules for `/` and `~/` because their return type is not `num`, and the return value's type is independent of the argument types. A `/` operation always returns a `double` and a `~/` operation always returns an `int`.
+
+Tabulated, the rules for binary operators are:
+
+
+| op1 \\ op2  | int     | int+    | double  | double+ | num     | num+    |
+| :---------: | ------- | ------- | ------- | ------- | ------- | ------- |
+|   **int**   | int     | int     | double  | double  | num     | num     |
+|  **int+**   | int+    | int+    | double  | double  | num     | num     |
+| **double**  | double  | double  | double  | double  | double  | double  |
+| **double+** | double+ | double+ | double+ | double+ | double+ | double+ |
+|   **num**   | num     | num     | double  | double  | num     | num     |
+|  **num+**   | num+    | num+    | double  | double  | num+    | num+    |
```

Comment:
ACK. With the rewrite above, the table would be:

| op1 \\ op2  | int     | int+    | double  | double+ | num     | num+    |
| :---------: | ------- | ------- | ------- | ------- | ------- | ------- |
|   **int**   | int     | int     | double  | double  | num     | num     |
|  **int+**   | int+    | int+    | double  | double  | num     | num     |
| **double**  | double  | double  | double  | double  | double  | double  |
| **double+** | double+ | double+ | double+ | double+ | double+ | double+ |
|   **num**   | num     | num     | double  | double  | num     | num     |
|  **num+**   | num+    | num+    | double  | double  | num    | num/num+    |

(Only change is the two last `num`'s in the final row which are no longer `num+`. The `num/num+` covers the case where the two operands have the *same* type which extends num.).

Reply:
I think these tables should be non-normative (and given that feature specs usually don't make that distinction very strictly, we should probably say it in words), because they involve the subtleties around `int+` and `int+` being, say, `X extends int` and `Y & int` and similar cases where it isn't just the same type. The remaining text should be complete anyway.